### ISSUE TITLE
Error exception handling

### ISSUE THREAD

**Issue by DominicFuMJ** [nus-cs2103-AY2324S2/forum#210]
I'm trying to do the A-MoreOOP but not sure what to put for the catch part of the exception. I made it so that the TaskList would take in an ArrayList<Task> and not sure what exception to put if storage.load() does not output ArrayList<Task>.
![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/119829202/4171c905-dd6d-4d7f-a256-4a0c9482ff78)

**Comment by damithc** [comment]
@DominicFuMJ the following can be a simple way to handle loading errors. It is 'simple' because it treats all possible loading errors in the same way (rather than respond to different types of errors in different ways) but it's good enough for a start.

```java
} catch (Exception e) {
    tasks = new TaskList();
}
```

**Comment by NeoHW** [comment]
If you are planning to catch specific errors, A method that I did is to do it without the try-catch block first, and intentionally feed it inputs which will give errors.
Then, I would be able to know what errors are there to catch the specific errors.

**Comment by alfaloo** [comment]
Another possible solution would be to write a custom exceptions class that extends ur main DukeExceptions. Then when u are unsure which exception to catch, you can just catch the main DukeException. This is not perfect because it is not specific, but it is far better than just the Pokemon catch with "catch (Exception e)". Hope this helps c:

**Comment by UdhayaShan1** [comment]
try catch imo shouldnt really be a control flow thing (correct me if im wrong) so I think its best to print the exception's message if caught and not do any operations?

If im not wrong, you are trying to load the filepath so check if exists and perform operations first and if exception still occurs raise this to the user.

**Comment by camille-readbean** [comment]
I think its depends on how your application is designed. Should `storage.load` _always_ return a tasklist?  
Should `storage.load` handle the errors within itself and _"fail gracefully"_ or should the error be passed up to `Victor` who decide what to do?  
For example within `storage.load` itself you can catch exceptions for File not existing, being unable to parse and load the Tasks or IO errors. 
  
Extending what @alfaloo said, for example `storage.load` can throw a custom subclass of DukeException, maybe like a `StorageLoadException` so you can 

```java
try {
    ....
} catch (StorageLoadException e) {
    ... // do something
} catch (InvalidParseException e) {
  ... // do something else 
} catch (Exception e) {
  ... // do something else
}
```

@UdhayaShan1 
👍 😄 
Good point! Exceptions for control flows is something from _CS2030S_ (PTSD lol) and I don't think handling __edge cases__ is that bad of a "pattern".   
I think in this case it is fine, the alternative is to also have guard conditions; to ensure file is there, to ensure file format is valid.  
But if we do this we run into a problem with how do we tell the calling function that something Succeeded or not and we may end up having to use Monads ala [`Results (rust)`](https://doc.rust-lang.org/std/result/) [`Results (kotlin)`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/)  
  
According to the textbook:  
  
> In general, use exceptions only for 'unusual' conditions. Use normal return statements to pass control to the caller for conditions that are 'normal'.  
https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/errorHandling.html#error-handling 
  
Edit: 
TL;DR  
Depends on your separation of concern, error boundaries and what failure should do / means and how you want your software to respond to it.

**Comment by camille-readbean** [comment]
See also:  
  
https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why  
  
https://web.archive.org/web/20140430044213/http://c2.com/cgi-bin/wiki?DontUseExceptionsForFlowControl

**Comment by DominicFuMJ** [comment]
Place the exception within the Storage, and remove the try catch for the main file since the Storage.load will now always return a ArrayList. Thanks for the help everyone
